### DISCUSSION TITLE
Matching differs slightly from OSM-hosted instance

### DISCUSSION THREAD

**Discussion by tpalts** [mediagis/nominatim-docker#289] [Q&A]
Hello!

For some addresses, the matching process between our local instance and the OSM-hosted instance seems to differ.

Simple example:
https://nominatim.openstreetmap.org/ui/search.html?street=VIA+NAZIONALE+DELLE+PUGLIE+KM+840&city=CASORIA&country=IT

The OSM-hosted instance returns several results, the first of which is osm_id = W908701685.

When I run a search on our instance with the exact same GET parameters, I get no results. However, running the details request locally confirms that this Node DOES exist in our DB: `GET /details?osmtype=W&osmid=908701685`.

Our local instance starts returning results when we wither remove the house number `840` from the `street` parameter, or when we remove the token `KM`.

What could be causing this inconsistency between the two instances?

Thanks for your help :) Awesome project, by the way 👏

Details:
- We are running v4.0 with the Europe data extract from https://download.geofabrik.de/

**Comment by mtmail** [comment]
In general 'km 840' is not supported and not data mapped in OpenStreetMap so it's always better to remove it from the query.

https://nominatim.openstreetmap.org uses the `master` branch of https://github.com/osm-search/nominatim and is ahead of the stable version 4.0.1 by two months.

The tokenizer got completely rewritten and https://github.com/osm-search/Nominatim/pull/2588 improves the house-number matching (not clear if that explains the difference for this query though as KM 840 is not a house number).

You can try the new tokenizer with the `NOMINATIM_TOKENIZER` variable https://nominatim.org/2021/11/03/release-40.html

I should note at the same time https://nominatim.openstreetmap.org is also behind, e.g. reindexing OSM data, so it can happen results on your local fresh installed instance are better in some regions of the world.

**Reply by tpalts** [reply]
Thanks a lot for your answer. I would like to try the new `icu` tokenizer that's mentioned in the release announcement. For this, the author suggests that the following be added to Nominatim's `.env`: `NOMINATIM_TOKENIZER=icu`

In the context of the repo at hand, is it enough to just add this as an environment variable in the docker command?

Example:
`docker run -it --rm --shm-size=64g  -e NOMINATIM_TOKENIZER=icu ...  mediagis/nominatim:4.0`

**Reply by mtmail** [reply]
Add the setting to your local `conf.d/env` and build a local image as fork. If I understand the `Dockerfile` correct that will end up as `.env` in Nominatim's project directory.

**Reply by leonardehrenfried** [reply]
@mtmail Does nominatim not pick up regular env variables? Just the stuff in `.env`?

If this is so, then we also would merge in a PR that makes this configurable, @tpalts.

**Reply by mtmail** [reply]
Nominatim uses `dotenv` and should pick up regular env variables as well https://nominatim.org/release-docs/latest/customize/Settings/#configuring-nominatim

**Reply by tpalts** [reply]
I finally managed to set up the instance with `NOMINATIM_TOKENIZER=icu`. However, the issue still remains - my instance provides no results for the above-mentioned search, while the OSM-hosted instance does.

What other explanations may there be for this discrepency?

**Reply by leonardehrenfried** [reply]
Are you using the latest master?

**Reply by tpalts** [reply]
Yes. Although some very minor tweaks were needed to make it run with the new tokenizer. See [my fork here](https://github.com/tpalts/nominatim-docker/commits/contribution).
